cmd/src/lsif_upload.go

```diff
@@ -219,18 +219,35 @@ func (e errorWithHint) Error() string {
 	return fmt.Sprintf("%s\n\n%s\n", e.err, e.hint)
 }
 
-var errUnauthorizedHint = strings.Join([]string{
-	"You may need to specify or update your GitHub access token to use this endpoint.",
-	"See https://docs.sourcegraph.com/cli/references/lsif/upload.",
-}, "\n")
-
 // handleLSIFUploadError writes the given error to the given output. If the
 // given output object is nil then the error will be written to standard out.
 //
 // This method returns the error that should be passed back up to the runner.
 func handleLSIFUploadError(out *output.Output, err error) error {
 	if err == upload.ErrUnauthorized {
-		err = errorWithHint{err: err, hint: errUnauthorizedHint}
+		var actionableHints []string
+		if lsifUploadFlags.gitHubToken != "" {
+			actionableHints = append(actionableHints,
+				"The supplied -github-token did not give the expected access level for the target repository.",
+				"Please check that supplied token matches the code host and has expected access of the target repository.",
+			)
+		} else if lsifUploadFlags.gitLabToken != "" {
+			actionableHints = append(actionableHints,
+				"The supplied -gitlab-token did not give the expected access level for the target repository.",
+				"Please check that supplied token matches the code host and has expected access of the target repository.",
+			)
+		} else {
+			actionableHints = append(actionableHints,
+				"Please retry your request with a -github-token or -gitlab-token (matching the code host of the target repository).",
```

Comment:
Can we pass in the repo URL so that we can emit a precise hint here about -github-token or -gitlab-token?

Reply:
We don't reach out to any other code host (because we can't supply any other tokens). We could do the [same check](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@a2c3de6ceb4a70c9ad9946797fd39c3fd6b2bf40/-/blob/enterprise/cmd/frontend/internal/codeintel/httpapi/auth_middleware.go?L25) we do in the frontend to figure out which code host it is, though.

Reply:
> We don't reach out to any other code host (because we can't supply any other tokens).

Hmm, I hadn't looked at the code with too much attention; I'd assumed this same code path would apply for other code hosts too, and that we don't currently support access tokens for other code hosts.

Doing some check for a more targeted message (particularly around the access scopes part below) sounds good. 👍

Reply:
We return a specific error in the non-github/non-gitlab case: https://sourcegraph.com/github.com/sourcegraph/sourcegraph@947423ea4f32fe4766d406b4ad1dff08407a1e20/-/blob/enterprise/cmd/frontend/internal/codeintel/httpapi/auth_middleware.go?L64